extensions/email/src/main/java/org/mule/extension/email/internal/commands/ListCommand.java

```diff
@@ -0,0 +1,85 @@
+import org.mule.extension.email.api.retriever.RetrieverConnection;
+import org.mule.extension.email.api.EmailAttributes;
+import org.mule.extension.email.internal.exception.EmailRetrieverException;
+import org.mule.runtime.api.message.MuleMessage;
+import org.mule.runtime.core.DefaultMuleMessage;
+import org.mule.runtime.core.api.MuleContext;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import javax.mail.Folder;
+import javax.mail.Message;
+import javax.mail.MessagingException;
+
+/**
+ * Represents the list emails operation.
+ *
+ * @since 4.0
+ */
+public final class ListCommand
+{
+
+    /**
+     * Retrieves all the emails in the specified {@code folderName}.
+     * <p>
+     * A new {@link MuleMessage} is created for each fetched email from the folder, where
+     * the payload is the text body of the email and the other metadata is carried by
+     * an {@link EmailAttributes} instance.
+     * <p>
+     * For folder implementations (like IMAP) that support fetching without reading the content, if the content
+     * should NOT be read ({@code readContent} = false) the SEEN flag is not going to be set.
+     *
+     * @param connection  the associated {@link RetrieverConnection}.
+     * @param context     the mule context to create the result message.
+     * @param folderName  the name of the folder where the emails are stored.
+     * @param readContent if should read the email content or not.
+     * @param matcher     a {@link Predicate} of {@link EmailAttributes} used to filter the output list  @return a {@link List} of {@link MuleMessage} carrying all the emails and it's corresponding attributes.
+     */
+    public List<MuleMessage<String, EmailAttributes>> list(RetrieverConnection connection,
+                                                           MuleContext context,
+                                                           String folderName,
+                                                           boolean readContent,
+                                                           Predicate<EmailAttributes> matcher)
+    {
+        Folder folder = getOpenFolder(folderName, READ_ONLY, connection.getStore());
+        try
+        {
+            List<MuleMessage<String, EmailAttributes>> list = new LinkedList<>();
+            for (Message m : folder.getMessages())
+            {
+                String body = "";
+                EmailAttributes attributes = fromMessage(m, false);
+                if (matcher.test(attributes))
+                {
+                    if (readContent)
+                    {
+                        body = getTextBody(m);
+                        attributes = fromMessage(m);
+                    }
+                    list.add((MuleMessage) new DefaultMuleMessage(body, null, attributes, context));
+                }
+            }
+            closeFolder(folder, false);
```

Comment:
This is still wrong. The folder should be closed on a finally block, and it shouldn't be up to the developer to remember to do so. This should be centralized on the connection provider through both, a pooling listener and the disconnect method. 

It's also not a good idea to have `getOpenFolder` as a static method. It should be part of the connection state, otherwise it's very difficult to test